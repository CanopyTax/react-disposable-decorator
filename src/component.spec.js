import { CancelableComponent } from "./react-cancelable.js";
import { Observable } from "rx";
import { mount } from "enzyme";
import React from "react";

describe("CancelableComponent", function() {
	it("should render", function() {
		function Component({ cancelWhenUnmounted, cancelAllSubscriptions, stuff }) {
			return <div>Hello {stuff}</div>;
		}

		const wrapper = mount(
			<CancelableComponent>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</CancelableComponent>
		);

		expect(wrapper).toMatchSnapshot();
	});

	it("should cancel subscriptions", function(done) {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					Observable.timer(100).subscribe(() => {
						fail('should have canceled the subscription');
					})
				);
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const wrapper = mount(
			<CancelableComponent>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</CancelableComponent>
		);

		wrapper.unmount();
		setTimeout(done, 200);
	});

	it("should cancel all subscriptions", function(done) {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					Observable.timer(100).subscribe(() => {
						fail('should have canceled the subscription');
					})
				);

				this.props.cancelWhenUnmounted(
					Observable.timer(50).subscribe(() => {
						fail('should have canceled the subscription');
					})
				);

				this.props.cancelAllSubscriptions();
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const wrapper = mount(
			<CancelableComponent>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</CancelableComponent>
		);

		setTimeout(done, 200);
	});

	it("should throw when a non subscription is passed", function() {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					"this should not work"
				);
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const makeError = () => mount(
			<CancelableComponent>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</CancelableComponent>
		);

		expect(makeError).toThrowError('cancelWhenUnmounted should be called with one or more disposables (an object with a dispose function)');
	});
});
