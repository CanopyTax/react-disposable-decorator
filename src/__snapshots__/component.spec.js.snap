// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CancelableComponent should render 1`] = `
<CancelableComponent>
  <Component
    cancelAllSubscriptions={[Function]}
    cancelWhenUnmounted={[Function]}
    stuff={1}
  >
    <div>
      Hello 
      1
    </div>
  </Component>
</CancelableComponent>
`;
